File: backend/src/controllers/authController.ts
import { Request, Response, NextFunction } from 'express';
import * as authService from '../services/authService';
import prisma from '../config/db';
import ErrorResponse from '../utils/errorResponse';
const sendTokenCookies = (user: any, accessToken: string, refreshToken: string, statusCode: number, res: Response) => {
  const accessCookieOptions = {
    expires: new Date(Date.now() + 15 * 60 * 1000),
    httpOnly: true,
    secure: process.env.NODE_ENV === 'production',
    sameSite: process.env.NODE_ENV === 'production' ? ('none' as const) : ('lax' as const)
  };
  const refreshCookieOptions = {
    expires: new Date(Date.now() + 7 * 24 * 60 * 60 * 1000),
    httpOnly: true,
    secure: process.env.NODE_ENV === 'production',
    sameSite: process.env.NODE_ENV === 'production' ? ('none' as const) : ('lax' as const)
  };
  res
    .status(statusCode)
    .cookie('token', accessToken, accessCookieOptions)
    .cookie('refreshToken', refreshToken, refreshCookieOptions)
    .json({
      success: true,
      user
    });
};
export const register = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { name, email, password } = req.body;
    await authService.registerUser(name, email, password);
    res.status(201).json({
      success: true,
      message: 'Registration successful! Please verify your email using the link printed to the server console.'
    });
  } catch (error) {
    next(error);
  }
};
export const login = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { email, password } = req.body;
    const userAgent = req.headers['user-agent'];
    const ipAddress = req.ip;
    const { user, accessToken, refreshToken } = await authService.loginUser(email, password, userAgent, ipAddress);
    sendTokenCookies(user, accessToken, refreshToken, 200, res);
  } catch (error) {
    next(error);
  }
};
export const logout = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const refreshToken = req.cookies.refreshToken;
    if (refreshToken) {
      await prisma.session.deleteMany({
        where: { refresh_token: refreshToken }
      });
    }
    res.cookie('token', 'none', {
      expires: new Date(Date.now() + 10 * 1000),
      httpOnly: true,
      secure: process.env.NODE_ENV === 'production',
      sameSite: process.env.NODE_ENV === 'production' ? ('none' as const) : ('lax' as const)
    });
    res.cookie('refreshToken', 'none', {
      expires: new Date(Date.now() + 10 * 1000),
      httpOnly: true,
      secure: process.env.NODE_ENV === 'production',
      sameSite: process.env.NODE_ENV === 'production' ? ('none' as const) : ('lax' as const)
    });
    res.status(200).json({
      success: true,
      message: 'Logged out successfully.'
    });
  } catch (error) {
    next(error);
  }
};
export const getMe = async (req: any, res: Response, next: NextFunction) => {
  try {
    res.status(200).json({
      success: true,
      user: req.user
    });
  } catch (error) {
    next(error);
  }
};
export const verify = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { token } = req.query;
    if (!token || typeof token !== 'string') {
      return res.redirect('http://localhost:5173/login?error=invalid_token');
    }
    const user = await prisma.user.findUnique({
      where: { verification_token: token }
    });
    if (!user) {
      return res.redirect('http://localhost:5173/login?error=invalid_token');
    }
    await prisma.user.update({
      where: { id: user.id },
      data: {
        is_verified: true,
        verification_token: null
      }
    });
    res.redirect('http://localhost:5173/login?verified=true');
  } catch (error) {
    next(error);
  }
};
export const refresh = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const refreshToken = req.cookies.refreshToken;
    if (!refreshToken) {
      throw new ErrorResponse('No refresh token provided. Please log in again.', 401);
    }
    const { accessToken } = await authService.refreshAccessToken(refreshToken);
    const accessCookieOptions = {
      expires: new Date(Date.now() + 15 * 60 * 1000),
      httpOnly: true,
      secure: process.env.NODE_ENV === 'production',
      sameSite: process.env.NODE_ENV === 'production' ? ('none' as const) : ('lax' as const)
    };
    res
      .status(200)
      .cookie('token', accessToken, accessCookieOptions)
      .json({
        success: true
      });
  } catch (error) {
    next(error);
  }
};
export const getSessions = async (req: any, res: Response, next: NextFunction) => {
  try {
    const sessions = await authService.getActiveSessions(req.user.id);
    const currentToken = req.cookies.refreshToken;
    const formattedSessions = sessions.map(session => ({
      id: session.id,
      device_name: session.device_name,
      ip_address: session.ip_address,
      created_at: session.created_at,
      is_current: session.refresh_token === currentToken
    }));
    res.status(200).json({
      success: true,
      sessions: formattedSessions
    });
  } catch (error) {
    next(error);
  }
};
export const deleteSession = async (req: any, res: Response, next: NextFunction) => {
  try {
    const { id } = req.params;
    await authService.terminateSession(req.user.id, id);
    res.status(200).json({
      success: true,
      message: 'Session terminated successfully.'
    });
  } catch (error) {
    next(error);
  }
};
export const logoutAll = async (req: any, res: Response, next: NextFunction) => {
  try {
    await authService.terminateAllSessions(req.user.id);
    res.cookie('token', 'none', {
      expires: new Date(Date.now() + 10 * 1000),
      httpOnly: true,
      secure: process.env.NODE_ENV === 'production',
      sameSite: process.env.NODE_ENV === 'production' ? ('none' as const) : ('lax' as const)
    });
    res.cookie('refreshToken', 'none', {
      expires: new Date(Date.now() + 10 * 1000),
      httpOnly: true,
      secure: process.env.NODE_ENV === 'production',
      sameSite: process.env.NODE_ENV === 'production' ? ('none' as const) : ('lax' as const)
    });
    res.status(200).json({
      success: true,
      message: 'Logged out from all devices successfully.'
    });
  } catch (error) {
    next(error);
  }
};
export const forgotPassword = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { email } = req.body;
    await authService.forgotPassword(email);
    res.status(200).json({
      success: true,
      message: 'If an account with that email exists, a password reset link has been sent.'
    });
  } catch (error) {
    next(error);
  }
};
export const resetPassword = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { token, newPassword } = req.body;
    await authService.resetPassword(token, newPassword);
    res.cookie('token', 'none', { expires: new Date(Date.now() + 10 * 1000), httpOnly: true });
    res.cookie('refreshToken', 'none', { expires: new Date(Date.now() + 10 * 1000), httpOnly: true });
    res.status(200).json({
      success: true,
      message: 'Password has been successfully reset. Please log in with your new password.'
    });
  } catch (error) {
    next(error);
  }
};
export const changePassword = async (req: any, res: Response, next: NextFunction) => {
  try {
    const { oldPassword, newPassword } = req.body;
    await authService.changePassword(req.user.id, oldPassword, newPassword);
    res.status(200).json({
      success: true,
      message: 'Password changed successfully.'
    });
  } catch (error) {
    next(error);
  }
};

File: backend/src/main/java/com/urlshortener/controller/AuthController.java
package com.urlshortener.controller;

import com.urlshortener.dto.request.ChangePasswordRequest;
import com.urlshortener.dto.request.ForgotPasswordRequest;
import com.urlshortener.dto.request.LoginRequest;
import com.urlshortener.dto.request.RegisterRequest;
import com.urlshortener.dto.request.ResetPasswordRequest;
import com.urlshortener.entity.User;
import com.urlshortener.exception.ApiException;
import com.urlshortener.repository.SessionRepository;
import com.urlshortener.repository.UserRepository;
import com.urlshortener.service.AuthService;

import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import jakarta.validation.Valid;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.net.URI;
import java.util.*;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/auth")
public class AuthController {

    @Autowired
    private AuthService authService;

    @Autowired
    private SessionRepository sessionRepository;

    @Autowired
    private UserRepository userRepository;

    @Value("${client.url}")
    private String clientUrl;

    private Cookie createCookie(String name, String value, int maxAge) {
        Cookie cookie = new Cookie(name, value);
        cookie.setHttpOnly(true);
        cookie.setPath("/");
        cookie.setSecure(false); 
        cookie.setMaxAge(maxAge);
        return cookie;
    }

    private void setTokenCookie(HttpServletResponse response, String accessToken) {
        response.addCookie(createCookie("token", accessToken, 15 * 60)); 
    }

    private void setRefreshTokenCookie(HttpServletResponse response, String refreshToken) {
        response.addCookie(createCookie("refreshToken", refreshToken, 7 * 24 * 60 * 60)); 
    }

    private void clearCookies(HttpServletResponse response) {
        response.addCookie(createCookie("token", "none", 10));
        response.addCookie(createCookie("refreshToken", "none", 10));
    }

    private String getCookieValue(HttpServletRequest request, String name) {
        if (request.getCookies() != null) {
            for (Cookie cookie : request.getCookies()) {
                if (cookie.getName().equals(name)) {
                    return cookie.getValue();
                }
            }
        }
        return null;
    }

    @PostMapping("/register")
    public ResponseEntity<?> register(@Valid @RequestBody RegisterRequest request) {
        Map<String, Object> userInfo = authService.registerUser(
                request.getName(), request.getEmail(), request.getPassword());

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("success", true);
        response.put("message", "Registration successful! Please check your email to verify your account.");
        response.put("user", userInfo);

        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    @PostMapping("/login")
    public ResponseEntity<?> login(@Valid @RequestBody LoginRequest request,
                                   HttpServletRequest httpRequest,
                                   HttpServletResponse httpResponse) {
        String userAgent = httpRequest.getHeader("User-Agent");
        String ipAddress = httpRequest.getRemoteAddr();

        Map<String, Object> result = authService.loginUser(
                request.getEmail(), request.getPassword(), userAgent, ipAddress);

        setTokenCookie(httpResponse, (String) result.get("accessToken"));
        setRefreshTokenCookie(httpResponse, (String) result.get("refreshToken"));

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("success", true);
        response.put("user", result.get("user"));

        return ResponseEntity.ok(response);
    }

    @PostMapping("/logout")
    public ResponseEntity<?> logout(HttpServletRequest request, HttpServletResponse response) {
        String refreshToken = getCookieValue(request, "refreshToken");

        if (refreshToken != null && !"none".equals(refreshToken)) {
            
            sessionRepository.findByRefreshToken(refreshToken)
                    .ifPresent(session -> sessionRepository.delete(session));
        }

        clearCookies(response);

        Map<String, Object> responseBody = new LinkedHashMap<>();
        responseBody.put("success", true);
        responseBody.put("message", "Logged out successfully.");

        return ResponseEntity.ok(responseBody);
    }

    @GetMapping("/me")
    public ResponseEntity<?> getMe(HttpServletRequest request) {
        @SuppressWarnings("unchecked")
        Map<String, Object> user = (Map<String, Object>) request.getAttribute("user");
        UUID userId = (UUID) request.getAttribute("userId");

        if (user == null) {
            return ResponseEntity.status(401).body(Map.of("success", false, "error", "Unauthorized"));
        }

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("success", true);
        response.put("user", user);

        return ResponseEntity.ok(response);
    }

    @GetMapping("/verify")
    public ResponseEntity<?> verifyEmail(@RequestParam("token") String token,
                                          HttpServletResponse response) {
        try {
            Optional<User> userOpt = userRepository.findByVerificationToken(token);

            if (userOpt.isEmpty()) {
                return ResponseEntity.status(HttpStatus.FOUND)
                        .location(URI.create(clientUrl + "/login?error=invalid_token"))
                        .build();
            }

            User user = userOpt.get();
            user.setVerified(true);
            user.setVerificationToken(null);
            userRepository.save(user);

            return ResponseEntity.status(HttpStatus.FOUND)
                    .location(URI.create(clientUrl + "/login?verified=true"))
                    .build();
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.FOUND)
                    .location(URI.create(clientUrl + "/login?error=invalid_token"))
                    .build();
        }
    }

    @PostMapping("/refresh")
    public ResponseEntity<?> refreshToken(HttpServletRequest request,
                                           HttpServletResponse httpResponse) {
        String refreshToken = getCookieValue(request, "refreshToken");

        if (refreshToken == null || "none".equals(refreshToken)) {
            throw new ApiException("No refresh token provided.", 401);
        }

        Map<String, Object> result = authService.refreshAccessToken(refreshToken);

        setTokenCookie(httpResponse, (String) result.get("accessToken"));

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("success", true);

        return ResponseEntity.ok(response);
    }

    @GetMapping("/sessions")
    public ResponseEntity<?> getSessions(HttpServletRequest request) {
        @SuppressWarnings("unchecked")
        Map<String, Object> user = (Map<String, Object>) request.getAttribute("user");
        UUID userId = (UUID) request.getAttribute("userId");

        if (user == null) {
            return ResponseEntity.status(401).body(Map.of("success", false, "error", "Unauthorized"));
        }

        String currentRefreshToken = getCookieValue(request, "refreshToken");
        List<Map<String, Object>> sessions = authService.getActiveSessions(userId);

        List<Map<String, Object>> sessionsWithCurrent = sessions.stream().map(session -> {
            Map<String, Object> sessionCopy = new LinkedHashMap<>(session);
            String sessionRefreshToken = (String) session.get("refresh_token");
            sessionCopy.put("is_current", sessionRefreshToken != null &&
                    sessionRefreshToken.equals(currentRefreshToken));
            sessionCopy.remove("refresh_token"); 
            return sessionCopy;
        }).collect(Collectors.toList());

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("success", true);
        response.put("sessions", sessionsWithCurrent);

        return ResponseEntity.ok(response);
    }

    @DeleteMapping("/sessions/{id}")
    public ResponseEntity<?> terminateSession(@PathVariable("id") String id,
                                               HttpServletRequest request) {
        @SuppressWarnings("unchecked")
        Map<String, Object> user = (Map<String, Object>) request.getAttribute("user");
        UUID userId = (UUID) request.getAttribute("userId");

        if (user == null) {
            return ResponseEntity.status(401).body(Map.of("success", false, "error", "Unauthorized"));
        }

        authService.terminateSession(userId, id);

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("success", true);
        response.put("message", "Session terminated successfully.");

        return ResponseEntity.ok(response);
    }

    @PostMapping("/logout-all")
    public ResponseEntity<?> logoutAll(HttpServletRequest request,
                                        HttpServletResponse httpResponse) {
        @SuppressWarnings("unchecked")
        Map<String, Object> user = (Map<String, Object>) request.getAttribute("user");
        UUID userId = (UUID) request.getAttribute("userId");

        if (user == null) {
            return ResponseEntity.status(401).body(Map.of("success", false, "error", "Unauthorized"));
        }

        authService.terminateAllSessions(userId);
        clearCookies(httpResponse);

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("success", true);
        response.put("message", "Logged out from all devices successfully.");

        return ResponseEntity.ok(response);
    }

    @PostMapping("/forgot-password")
    public ResponseEntity<?> forgotPassword(@Valid @RequestBody ForgotPasswordRequest request) {
        try {
            authService.forgotPassword(request.getEmail());
        } catch (ApiException e) {

        }

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("success", true);
        response.put("message", "If an account with that email exists, a password reset link has been sent.");

        return ResponseEntity.ok(response);
    }

    @PostMapping("/reset-password")
    public ResponseEntity<?> resetPassword(@Valid @RequestBody ResetPasswordRequest request,
                                            HttpServletResponse httpResponse) {
        authService.resetPassword(request.getToken(), request.getNewPassword());
        clearCookies(httpResponse);

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("success", true);
        response.put("message", "Password has been successfully reset. You can now log in with your new password.");

        return ResponseEntity.ok(response);
    }

    @PostMapping("/change-password")
    public ResponseEntity<?> changePassword(@Valid @RequestBody ChangePasswordRequest request,
                                             HttpServletRequest httpRequest) {
        @SuppressWarnings("unchecked")
        Map<String, Object> user = (Map<String, Object>) httpRequest.getAttribute("user");
        UUID userId = (UUID) httpRequest.getAttribute("userId");

        if (user == null) {
            return ResponseEntity.status(401).body(Map.of("success", false, "error", "Unauthorized"));
        }

        authService.changePassword(userId, request.getOldPassword(), request.getNewPassword());

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("success", true);
        response.put("message", "Password changed successfully.");

        return ResponseEntity.ok(response);
    }
}
